{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pyquery import PyQuery as pq\n",
    "import numpy as np\n",
    "import re\n",
    "import requests\n",
    "\n",
    "from pattern.en import parse\n",
    "from pattern.en import pprint\n",
    "from pattern.vector import stem, PORTER, LEMMA\n",
    "from sklearn.feature_extraction import text \n",
    "\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.feature_extraction.text import TfidfTransformer\n",
    "from gensim.models.ldamodel import LdaModel\n",
    "\n",
    "import os\n",
    "import json\n",
    "\n",
    "# Classifiers\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "import sklearn.linear_model as lm\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.svm import SVC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Text Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###########\n",
    "### Helper for text scrapping\n",
    "###########\n",
    "\n",
    "# Extract from the raw wikipedia page the text content\n",
    "# of the paragraphs and the table of content as list of words\n",
    "def get_text(data):\n",
    "    # Extracting text content in paragraphs\n",
    "    d_ = pq(data)\n",
    "    d_p = pq(d_('#mw-content-text p'))\n",
    "    paragraph = ''\n",
    "    for r in d_p:\n",
    "        paragraph += ' '+pq(r).text()\n",
    "\n",
    "    # Extracing Table of Contents\n",
    "    toc = []\n",
    "    d_table = pq(d_('#toc li'))\n",
    "    for c in d_table:\n",
    "        t = pq(c).text()\n",
    "        toc.append(' '.join(t.split(' ')[1:]))\n",
    "\n",
    "    return paragraph, toc\n",
    "\n",
    "# Return the title of the article, will be used to name the disease\n",
    "def get_title(data):\n",
    "    return pq(data)('#firstHeading').text()\n",
    "\n",
    "# Check if string contains digit\n",
    "def contains_digits(d):\n",
    "    _digits = re.compile('\\d')\n",
    "    return bool(_digits.search(d))\n",
    "\n",
    "# Return the common nouns in text (preprocessed) as a concatenated string\n",
    "def get_words(thetext):\n",
    "    stopwords=text.ENGLISH_STOP_WORDS\n",
    "    punctuation = list('.,;:!?()[]{}`''\\\"@#$^&*+-|=~_')\n",
    "    nouns = ''\n",
    "    descriptives = ''\n",
    "    proper = ''\n",
    "    sentences = parse(thetext, tokenize=True, lemmata=True).split()\n",
    "    for s in sentences:\n",
    "        for token in s:\n",
    "            if len(token[4]) > 1 and token[4] not in stopwords and token[4][0] not in punctuation and not contains_digits(token[4]):\n",
    "                if token[1] in ['JJ', 'JJR', 'JJS']:\n",
    "                    descriptives += ' '+ token[4]\n",
    "                elif token[1] in ['NN', 'NNS']:\n",
    "                    nouns += ' '+ token[4]\n",
    "                elif token[1] in ['NNP', 'NNPS']:\n",
    "                    proper += ' '+ token[4]\n",
    "    return nouns, descriptives, proper\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Script to retrieve and process the training text from the raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 iterations executed\n",
      "0 elements in the list\n",
      "100 iterations executed\n",
      "100 elements in the list\n",
      "200 iterations executed\n",
      "200 elements in the list\n",
      "300 iterations executed\n",
      "300 elements in the list\n",
      "400 iterations executed\n",
      "400 elements in the list\n",
      "500 iterations executed\n",
      "500 elements in the list\n",
      "600 iterations executed"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Script to retrieve the information from the training set\n",
    "for class_ in ['positive', 'negative']:\n",
    "    path = 'training/{}'.format(class_)\n",
    "    # List of considered files name\n",
    "    files = [f for f in os.listdir(\"./\"+path)]\n",
    "    # File not to be considered\n",
    "    bug = 'all-urls.txt'\n",
    "    if bug in files:\n",
    "        files.remove(bug)\n",
    "\n",
    "    # Initializiation\n",
    "    nouns_data = []\n",
    "    descriptives_data = []\n",
    "    proper_data = []\n",
    "    toc_data = []\n",
    "    titles = []\n",
    "\n",
    "    # Time indicator: \n",
    "    #     19.3s for 100 files in positive\n",
    "    #     4.43s for 100 files in negative\n",
    "    for i, filename in enumerate(files):\n",
    "        if i%100 == 0:\n",
    "            print '{} iterations executed'.format(i)\n",
    "            print '{} elements in the list'.format(len(nouns_data))\n",
    "        with open(path + '/' + filename) as f:\n",
    "            data = f.read()\n",
    "        # Titles\n",
    "        t = get_title(data)\n",
    "        titles.append(t)\n",
    "        # Content\n",
    "        paragraph, toc = get_text(data)\n",
    "        n, d, p = get_words(paragraph)\n",
    "        nouns_data.append(n)\n",
    "        descriptives_data.append(d)\n",
    "        proper_data.append(p)\n",
    "        toc_data.append(toc)\n",
    "\n",
    "    # Saving the outputs\n",
    "    with open('data/filesname_{}.json'.format(class_), 'w') as f:\n",
    "        json.dump(files, f)\n",
    "    with open('data/noun_list_{}.json'.format(class_), 'w') as f:\n",
    "        json.dump(nouns_data, f)\n",
    "    with open('data/descriptives_list_{}.json'.format(class_), 'w') as f:\n",
    "        json.dump(descriptives_data, f)\n",
    "    with open('data/proper_list_{}.json'.format(class_), 'w') as f:\n",
    "        json.dump(proper_data, f)\n",
    "    with open('data/toc_{}.json'.format(class_), 'w') as f:\n",
    "        json.dump(toc_data, f)\n",
    "    with open('data/title_list_{}.json'.format(class_), 'w') as f:\n",
    "        json.dump(titles, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classification"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploratory analysis of different classifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###########\n",
    "### Helper for feature extraction and model learning\n",
    "###########\n",
    "\n",
    "# Train the lda model to extract feature in the list of nouns in noun list.\n",
    "# Return the trained model (tuple (lda_ge, count_vect)) and the document_topics matrix from the list of nouns.\n",
    "def train_lda_model(noun_list, num_topics):\n",
    "    count_vect = CountVectorizer()\n",
    "    X_counts = count_vect.fit_transform(noun_list)\n",
    "    id2word = {k: v for v, k in count_vect.vocabulary_.iteritems()}\n",
    "    corpus = build_corpus(X_counts)\n",
    "    lda_ge = LdaModel(corpus, num_topics=num_topics, id2word=id2word)\n",
    "\n",
    "    # Building the topics distribution for each document\n",
    "    document_topics = np.zeros((len(corpus), num_topics))\n",
    "    for i, bow in enumerate(corpus):\n",
    "        dt = lda_ge.get_document_topics(bow)\n",
    "        for t in dt:\n",
    "            document_topics[i, t[0]] = t[1]\n",
    "\n",
    "    return document_topics, lda_ge, count_vect\n",
    "\n",
    "\n",
    "# Method to build the lda features for the document in the list of word\n",
    "# counts vector X_counts based on a trained lda model.\n",
    "def get_lda_features(X_counts, lda_ge):\n",
    "    corpus = build_corpus(X_counts)\n",
    "    infer = lda_ge.inference(corpus)[0]\n",
    "    # Need to normalize the gamma to have the topic distribution\n",
    "    document_topics = infer / np.sum(infer)\n",
    "\n",
    "    return document_topics\n",
    "\n",
    "\n",
    "# Build the corpus matrix in the required format to apply the lda\n",
    "def build_corpus(X_counts):\n",
    "    index = np.arange(X_counts.shape[1]).reshape((1, X_counts.shape[1]))\n",
    "    corpus = []\n",
    "    for i in xrange(X_counts.shape[0]):\n",
    "        if (X_counts[i, :] > 0).sum():\n",
    "            corpus.append([(w, c) for w, c in zip(index[np.where(\n",
    "                X_counts[i, :].toarray())],\n",
    "                np.array(X_counts[i, :][X_counts[i, :] > 0])[0])])\n",
    "        else:\n",
    "            corpus.append([])\n",
    "    return corpus\n",
    "\n",
    "\n",
    "# Function to build the occurence features and the pre-processing models from the list of nouns\n",
    "def get_features(noun_list):\n",
    "    # Building features\n",
    "    count_vect = CountVectorizer()\n",
    "    X_counts = count_vect.fit_transform(noun_list)\n",
    "    tf_transformer = TfidfTransformer(use_idf=False).fit(X_counts)\n",
    "    X_tf = tf_transformer.transform(X_counts)\n",
    "    \n",
    "    return X_tf, count_vect, tf_transformer\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Loading the pre_processed training data\n",
    "\n",
    "# Nouns\n",
    "with open('data/noun_list_negative.json', 'r') as f:\n",
    "    list_negative1 = json.load(f)\n",
    "with open('data/noun_list_positive.json', 'r') as f:\n",
    "    list_positive1 = json.load(f)\n",
    "\n",
    "# Descriptives\n",
    "with open('data/descriptives_list_negative.json', 'r') as f:\n",
    "    list_negative2 = json.load(f)\n",
    "with open('data/descriptives_list_positive.json', 'r') as f:\n",
    "    list_positive2 = json.load(f)\n",
    "\n",
    "# Filesname\n",
    "with open('data/filesname_positive.json', 'r') as f:\n",
    "    files_positive = json.load(f)\n",
    "with open('data/filesname_negative.json', 'r') as f:\n",
    "    files_negatives = json.load(f)\n",
    "    \n",
    "num_negative = len(files_negatives)\n",
    "num_positive = len(files_positive)\n",
    "\n",
    "list_negative = [ u + v for u,v in zip(list_negative1, list_negative2)]\n",
    "list_positive = [ u + v for u,v in zip(list_positive1, list_positive2)]\n",
    "\n",
    "noun_list = list_negative + list_positive\n",
    "filesname = files_negatives + files_positive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9583\n",
      "9583\n"
     ]
    }
   ],
   "source": [
    "# Building the mask, ie splitting among train and test set\n",
    "\n",
    "prop_train = 0.7\n",
    "\n",
    "# Negative\n",
    "mask_negative = np.zeros(num_negative, dtype=bool)\n",
    "mask_negative[:int(prop_train * num_negative)] = True\n",
    "np.random.shuffle(mask_negative)\n",
    "\n",
    "# Positive\n",
    "mask_positive = np.zeros(num_positive, dtype=bool)\n",
    "mask_positive[:int(prop_train * (num_positive))] = True\n",
    "np.random.shuffle(mask_positive)\n",
    "\n",
    "mask = np.concatenate((mask_negative, mask_positive))\n",
    "print(sum(mask))\n",
    "print(int(prop_train * (num_negative + num_positive)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LR: score on train 0.989669205885\n",
      "LR: score on test 0.984907497566\n",
      "LR: score on train 0.972242512783\n",
      "LR: score on test 0.97541382668\n",
      "stacked regression: score on train 0.990190963164\n",
      "stacked regression: score on test 0.985637779942\n",
      "CPU times: user 1min 13s, sys: 1.85 s, total: 1min 15s\n",
      "Wall time: 1min 18s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Stacked regression model\n",
    "\n",
    "# Variables:\n",
    "# noun_list, num_topics, mask\n",
    "num_topics = 60\n",
    "\n",
    "# Building the output column\n",
    "Y = np.concatenate((np.zeros(num_negative), np.ones(num_positive)))\n",
    "\n",
    "# Baseline\n",
    "baseline_acc_train = accuracy_score(Y[np.where(mask)], np.zeros(sum(mask)), normalize=True, sample_weight=None)\n",
    "baseline_acc_test = accuracy_score(Y[np.where(~mask)], np.zeros(sum(~mask)), normalize=True, sample_weight=None)\n",
    "print('Baseline (All negative): score on train {}'.format(baseline_acc_train))\n",
    "print('Baseline (All negative): score on test {}'.format(baseline_acc_test))\n",
    "\n",
    "# LR on Frequency matrix\n",
    "X_tf, count_vect_descriptive, tf_transformer_descriptive = get_features(noun_list)\n",
    "\n",
    "lr_occ = lm.LogisticRegression(penalty='l2')\n",
    "lr_occ.fit(X_tf[np.where(mask)], Y[np.where(mask)])\n",
    "print 'LR occ: score on train {}'.format(lr_occ.score(X_tf[np.where(mask)], Y[np.where(mask)]))\n",
    "print 'LR occ: score on test {}'.format(lr_occ.score(X_tf[np.where(~mask)], Y[np.where(~mask)]))\n",
    "\n",
    "Y_pred_occ = np.zeros((len(Y), 1))\n",
    "Y_pred_occ[mask, 0] = lr_occ.predict_proba(X_tf[np.where(mask)])[:, 0]\n",
    "Y_pred_occ[~mask, 0] = lr_occ.predict_proba(X_tf[np.where(~mask)])[:, 0]\n",
    "\n",
    "# LR on LDA features\n",
    "document_topics, lda_ge, count_vect = train_lda_model(noun_list, num_topics)\n",
    "\n",
    "lr_lda = lm.LogisticRegression(penalty='l2')\n",
    "lr_lda.fit(document_topics[np.where(mask)], Y[np.where(mask)])\n",
    "print 'LR lda: score on train {}'.format(lr_lda.score(document_topics[np.where(mask)], Y[np.where(mask)]))\n",
    "print 'LR lda: score on test {}'.format(lr_lda.score(document_topics[np.where(~mask)], Y[np.where(~mask)]))\n",
    "\n",
    "Y_pred_lda = np.zeros((len(Y), 1))\n",
    "Y_pred_lda[mask, 0] = lr_lda.predict_proba(document_topics[np.where(mask)])[:, 0]\n",
    "Y_pred_lda[~mask, 0] = lr_lda.predict_proba(document_topics[np.where(~mask)])[:, 0]\n",
    "\n",
    "# LR on the result\n",
    "\n",
    "X_lr = np.concatenate((Y_pred_occ, Y_pred_lda), axis=1)\n",
    "\n",
    "est = lm.LogisticRegression(penalty='l2')\n",
    "est.fit(X_lr[mask], Y[mask])\n",
    "print 'stacked regression: score on train {}'.format(est.score(X_lr[np.where(mask)], Y[np.where(mask)]))\n",
    "print 'stacked regression: score on test {}'.format(est.score(X_lr[np.where(~mask)], Y[np.where(~mask)]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SVM occ: score on train 0.998330376709\n",
      "SVM occ: score on test 0.990506329114"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SVM lda: score on train 0.976833976834\n",
      "SVM lda: score on test 0.979795520935\n",
      "stacked svm: score on train 0.998643431076\n",
      "stacked svm: score on test 0.989776046738\n",
      "CPU times: user 2min 11s, sys: 2.11 s, total: 2min 13s\n",
      "Wall time: 2min 15s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Stacked SVM model\n",
    "\n",
    "# Variables:\n",
    "# noun_list, num_topics, mask\n",
    "num_topics = 50\n",
    "\n",
    "# Building the output column\n",
    "Y = np.concatenate((np.zeros(num_negative), np.ones(num_positive)))\n",
    "\n",
    "# Baseline\n",
    "baseline_acc_train = accuracy_score(Y[np.where(mask)], np.zeros(sum(mask)), normalize=True, sample_weight=None)\n",
    "baseline_acc_test = accuracy_score(Y[np.where(~mask)], np.zeros(sum(~mask)), normalize=True, sample_weight=None)\n",
    "print('Baseline (All negative): score on train {}'.format(baseline_acc_train))\n",
    "print('Baseline (All negative): score on test {}'.format(baseline_acc_test))\n",
    "\n",
    "# SVM on Frequency matrix\n",
    "X_tf, count_vect, tf_transformer = get_features(noun_list)\n",
    "\n",
    "svm_occ = SVC(kernel = 'linear', probability=True)\n",
    "svm_occ.fit(X_tf[np.where(mask)], Y[np.where(mask)])\n",
    "print 'SVM occ: score on train {}'.format(svm_occ.score(X_tf[np.where(mask)], Y[np.where(mask)]))\n",
    "print 'SVM occ: score on test {}'.format(svm_occ.score(X_tf[np.where(~mask)], Y[np.where(~mask)]))\n",
    "\n",
    "Y_pred_occ = np.zeros((len(Y), 1))\n",
    "Y_pred_occ[mask, 0] = svm_occ.predict_proba(X_tf[np.where(mask)])[:, 0]\n",
    "Y_pred_occ[~mask, 0] = svm_occ.predict_proba(X_tf[np.where(~mask)])[:, 0]\n",
    "\n",
    "# SVM on LDA features\n",
    "document_topics, lda_ge, count_vect = train_lda_model(noun_list, num_topics)\n",
    "\n",
    "svm_lda = SVC(kernel = 'linear', probability=True)\n",
    "svm_lda.fit(document_topics[np.where(mask)], Y[np.where(mask)])\n",
    "print 'SVM lda: score on train {}'.format(svm_lda.score(document_topics[np.where(mask)], Y[np.where(mask)]))\n",
    "print 'SVM lda: score on test {}'.format(svm_lda.score(document_topics[np.where(~mask)], Y[np.where(~mask)]))\n",
    "\n",
    "Y_pred_lda = np.zeros((len(Y), 1))\n",
    "Y_pred_lda[mask, 0] = svm_lda.predict_proba(document_topics[np.where(mask)])[:, 0]\n",
    "Y_pred_lda[~mask, 0] = svm_lda.predict_proba(document_topics[np.where(~mask)])[:, 0]\n",
    "\n",
    "# SVM on the result\n",
    "\n",
    "X_lr = np.concatenate((Y_pred_occ, Y_pred_lda), axis=1)\n",
    "\n",
    "svm_stacked = SVC(kernel = 'linear', probability=True)\n",
    "svm_stacked.fit(X_lr[mask], Y[mask])\n",
    "print 'stacked svm: score on train {}'.format(svm_stacked.score(X_lr[np.where(mask)], Y[np.where(mask)]))\n",
    "print 'stacked svm: score on test {}'.format(svm_stacked.score(X_lr[np.where(~mask)], Y[np.where(~mask)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Baseline (All negative): score on train 0.73046018992\n",
      "Baseline (All negative): score on test 0.730282375852\n",
      "NB: score on train 0.98737347386\n",
      "NB: score on test 0.985637779942\n",
      "LR: score on train 0.988730042784\n",
      "LR: score on test 0.985150925024\n",
      "svm: score on train 0.996556401962\n",
      "svm: score on test 0.99123661149\n"
     ]
    }
   ],
   "source": [
    "# Set of classifiers tested\n",
    "\n",
    "# Baseline\n",
    "baseline_acc_train = accuracy_score(Y_train[np.where(mask)], np.zeros(sum(mask)), normalize=True, sample_weight=None)\n",
    "baseline_acc_test = accuracy_score(Y_train[np.where(~mask)], np.zeros(sum(~mask)), normalize=True, sample_weight=None)\n",
    "print('Baseline (All negative): score on train {}'.format(baseline_acc_train))\n",
    "print('Baseline (All negative): score on test {}'.format(baseline_acc_test))\n",
    "\n",
    "# Naive Bayes\n",
    "nb = MultinomialNB()\n",
    "nb.fit(X_train_tf[np.where(mask)], Y_train[np.where(mask)])\n",
    "print 'NB: score on train {}'.format(nb.score(X_tf[np.where(mask)], Y_train[np.where(mask)]))\n",
    "print 'NB: score on test {}'.format(nb.score(X_tf[np.where(~mask)], Y_train[np.where(~mask)]))\n",
    "\n",
    "# Logistic Regression\n",
    "lr = lm.LogisticRegression(penalty='l2')\n",
    "lr.fit(X_train_tf[np.where(mask)], Y_train[np.where(mask)])\n",
    "print 'LR: score on train {}'.format(lr.score(X_tf[np.where(mask)], Y_train[np.where(mask)]))\n",
    "print 'LR: score on test {}'.format(lr.score(X_tf[np.where(~mask)], Y_train[np.where(~mask)]))\n",
    "\n",
    "# SVM\n",
    "svc = SVC(kernel = 'linear')\n",
    "svc.fit(X_train_tf[np.where(mask)], Y_train[np.where(mask)])\n",
    "print 'svm: score on train {}'.format(svc.score(X_tf[np.where(mask)], Y_train[np.where(mask)]))\n",
    "print 'svm: score on test {}'.format(svc.score(X_tf[np.where(~mask)], Y_train[np.where(~mask)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LR: score on train 0.951476573098\n",
      "LR: score on test 0.95959104187 with 20"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LR: score on train 0.970572889492\n",
      "LR: score on test 0.972492697176 with 30"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LR: score on train 0.97182510696\n",
      "LR: score on test 0.968111002921 with 40"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LR: score on train 0.966085776897\n",
      "LR: score on test 0.974683544304 with 50"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LR: score on train 0.97036418658\n",
      "LR: score on test 0.974926971762 with 60"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:gensim.models.ldamodel:too few updates, training might not converge; consider increasing the number of passes or iterations to improve accuracy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LR: score on train 0.969946780758\n",
      "LR: score on test 0.967867575463 with 70\n"
     ]
    }
   ],
   "source": [
    "# number of topics tuning\n",
    "acc_test = []\n",
    "# LR on LDA features\n",
    "for num_topics in [20, 30, 40, 50, 60, 70]:\n",
    "    document_topics, lda_ge, count_vect = train_lda_model(noun_list, num_topics)\n",
    "\n",
    "    lr_lda_eval = SVC(kernel = 'linear')\n",
    "    lr_lda_eval.fit(document_topics[np.where(mask)], Y[np.where(mask)])\n",
    "    print 'LR: score on train {}'.format(lr_lda_eval.score(document_topics[np.where(mask)], Y[np.where(mask)]))\n",
    "    sc = lr_lda_eval.score(document_topics[np.where(~mask)], Y[np.where(~mask)])\n",
    "    print 'LR: score on test {} with {}'.format(sc, num_topics)\n",
    "    acc_test.append(sc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pipeline to build the selected classifier (called v1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Baseline (All negative): score on train 0.730406836608\n",
      "SVM occ: score on train 0.997516616756\n",
      "CPU times: user 1min 17s, sys: 894 ms, total: 1min 18s\n",
      "Wall time: 1min 22s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Loading the pre_processed training data\n",
    "\n",
    "# Nouns\n",
    "with open('data/noun_list_negative.json', 'r') as f:\n",
    "    list_negative1 = json.load(f)\n",
    "with open('data/noun_list_positive.json', 'r') as f:\n",
    "    list_positive1 = json.load(f)\n",
    "\n",
    "# Descriptives\n",
    "with open('data/descriptives_list_negative.json', 'r') as f:\n",
    "    list_negative2 = json.load(f)\n",
    "with open('data/descriptives_list_positive.json', 'r') as f:\n",
    "    list_positive2 = json.load(f)\n",
    "\n",
    "# Filesname\n",
    "with open('data/filesname_positive.json', 'r') as f:\n",
    "    files_positive = json.load(f)\n",
    "with open('data/filesname_negative.json', 'r') as f:\n",
    "    files_negatives = json.load(f)\n",
    "    \n",
    "num_negative = len(files_negatives)\n",
    "num_positive = len(files_positive)\n",
    "\n",
    "list_negative = [ u + v for u,v in zip(list_negative1, list_negative2)]\n",
    "list_positive = [ u + v for u,v in zip(list_positive1, list_positive2)]\n",
    "\n",
    "noun_list = list_negative + list_positive\n",
    "filesname = files_negatives + files_positive\n",
    "\n",
    "# Occurence SVM model training (on the whole dataset)\n",
    "\n",
    "# Variables:\n",
    "# noun_list, num_topics, mask\n",
    "num_topics = 50\n",
    "\n",
    "# Building the output column\n",
    "Y = np.concatenate((np.zeros(num_negative), np.ones(num_positive)))\n",
    "\n",
    "# Baseline\n",
    "baseline_acc_train = accuracy_score(Y, np.zeros(Y.shape[0]), normalize=True, sample_weight=None)\n",
    "print('Baseline (All negative): score on train {}'.format(baseline_acc_train))\n",
    "\n",
    "# SVM on Frequency matrix\n",
    "X_tf, count_vect, tf_transformer = get_features(noun_list)\n",
    "\n",
    "svm_occ = SVC(kernel = 'linear', probability=True)\n",
    "svm_occ.fit(X_tf, Y)\n",
    "print 'SVM occ: score on train {}'.format(svm_occ.score(X_tf, Y))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['model_v1/tf_transformer.pkl']"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Saving the pre-trained occurence svm model\n",
    "from sklearn.externals import joblib\n",
    "\n",
    "# Saving models\n",
    "joblib.dump(svm_occ, 'model_v1/svm_occ.pkl')\n",
    "joblib.dump(count_vect, 'model_v1/count_vect.pkl')\n",
    "joblib.dump(tf_transformer, 'model_v1/tf_transformer.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Improving the classifier for drugs: expanding the training set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scrapping the training set expansion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Loading the urls of drug pages\n",
    "drugs_url = []\n",
    "\n",
    "# List of drugs 1\n",
    "source = 'https://en.wikipedia.org/wiki/Anesthetic'\n",
    "t = requests.get(source)\n",
    "d = pq(t.text)('#mw-content-text')\n",
    "\n",
    "lis = d('.navbox-list li').remove('.new')\n",
    "for e in lis:\n",
    "    href = pq(e)('a').attr('href')\n",
    "    if href is not None:\n",
    "        drugs_url.append(href)\n",
    "\n",
    "# List of drugs 2\n",
    "source = 'https://en.wikipedia.org/wiki/List_of_antiviral_drugs'\n",
    "t = requests.get(source)\n",
    "d = pq(t.text)('#mw-content-text')\n",
    "\n",
    "lis = d.remove('div#toc')('li').remove('.new')\n",
    "for e in lis:\n",
    "    href = pq(e)('a').attr('href')\n",
    "    if href is not None:\n",
    "        drugs_url.append(href)\n",
    "\n",
    "# List of drugs 3\n",
    "source = 'https://en.wikipedia.org/wiki/List_of_largest_selling_pharmaceutical_products'\n",
    "t = requests.get(source)\n",
    "d = pq(t.text)('#mw-content-text')\n",
    "lis = d('table')('tr')\n",
    "\n",
    "for tr in lis:\n",
    "    href = pq(tr).children().eq(2)('a').attr('href')\n",
    "    if href is not None:\n",
    "        drugs_url.append(href)\n",
    "\n",
    "# List of drugs 4\n",
    "source = 'https://en.wikipedia.org/wiki/List_of_therapeutic_monoclonal_antibodies'\n",
    "t = requests.get(source)\n",
    "d = pq(t.text)('#mw-content-text')\n",
    "lis = d('table')('tr')\n",
    "\n",
    "for tr in lis:\n",
    "    href = pq(tr).children().eq(0).remove('.new').remove('.reference')('a').attr('href')\n",
    "    if href is not None:\n",
    "        drugs_url.append(href)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 iterations executed\n",
      "0 elements in the list\n",
      "100 iterations executed\n",
      "100 elements in the list\n",
      "200 iterations executed\n",
      "200 elements in the list\n",
      "300 iterations executed\n",
      "300 elements in the list\n",
      "400 iterations executed\n",
      "400 elements in the list\n",
      "500 iterations executed\n",
      "500 elements in the list\n",
      "600 iterations executed\n",
      "600 elements in the list\n",
      "700 iterations executed\n",
      "700 elements in the list\n"
     ]
    }
   ],
   "source": [
    "# Removing the duplicates\n",
    "files = list(set(drugs_url))\n",
    "\n",
    "# Script to extract noun list from drug pages\n",
    "wiki_url = 'https://en.wikipedia.org'\n",
    "class_ = 'drugs'\n",
    "\n",
    "# Initializiation\n",
    "nouns_data = []\n",
    "descriptives_data = []\n",
    "proper_data = []\n",
    "toc_data = []\n",
    "titles = []\n",
    "\n",
    "# Time indicator: \n",
    "#     19.3s for 100 files in positive\n",
    "#     4.43s for 100 files in negative\n",
    "for i, url in enumerate(files):\n",
    "    if i%100 == 0:\n",
    "        print '{} iterations executed'.format(i)\n",
    "        print '{} elements in the list'.format(len(nouns_data))\n",
    "    url_complete = wiki_url + url\n",
    "    t = requests.get(url_complete)\n",
    "    data = t.text\n",
    "    # Titles\n",
    "    title = get_title(data)\n",
    "    titles.append(title)\n",
    "    # Content\n",
    "    paragraph, toc = get_text(data)\n",
    "    n, d, p = get_words(paragraph)\n",
    "    nouns_data.append(n)\n",
    "    descriptives_data.append(d)\n",
    "    proper_data.append(p)\n",
    "    toc_data.append(toc)\n",
    "\n",
    "# Saving the outputs\n",
    "with open('data/filesname_{}.json'.format(class_), 'w') as f:\n",
    "    json.dump(files, f)\n",
    "with open('data/noun_list_{}.json'.format(class_), 'w') as f:\n",
    "    json.dump(nouns_data, f)\n",
    "with open('data/descriptives_list_{}.json'.format(class_), 'w') as f:\n",
    "    json.dump(descriptives_data, f)\n",
    "with open('data/proper_list_{}.json'.format(class_), 'w') as f:\n",
    "    json.dump(proper_data, f)\n",
    "with open('data/toc_{}.json'.format(class_), 'w') as f:\n",
    "    json.dump(toc_data, f)\n",
    "with open('data/title_list_{}.json'.format(class_), 'w') as f:\n",
    "    json.dump(titles, f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Building the new model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### VALIDATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10079\n",
      "10079\n",
      "Baseline (All negative): score on train 0.743724575851\n",
      "Baseline (All negative): score on test 0.743577875492\n",
      "SVM occ: score on train 0.99573370374\n",
      "SVM occ: score on test 0.987040037028\n",
      "CPU times: user 1min 13s, sys: 1.28 s, total: 1min 15s\n",
      "Wall time: 1min 19s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "### Training a new model with negative training set = negative + drugs \n",
    "# with a test set (to validate the accuracy)\n",
    "\n",
    "# Loading the pre_processed training data\n",
    "# Nouns\n",
    "with open('data/noun_list_negative.json', 'r') as f:\n",
    "    list_negative11 = json.load(f)\n",
    "with open('data/noun_list_drugs.json', 'r') as f:\n",
    "    list_negative12 = json.load(f)\n",
    "list_negative1 = list_negative11 + list_negative12\n",
    "with open('data/noun_list_positive.json', 'r') as f:\n",
    "    list_positive1 = json.load(f)\n",
    "\n",
    "# Descriptives\n",
    "with open('data/descriptives_list_negative.json', 'r') as f:\n",
    "    list_negative21 = json.load(f)\n",
    "with open('data/descriptives_list_drugs.json', 'r') as f:\n",
    "    list_negative22 = json.load(f)\n",
    "list_negative2 = list_negative21 + list_negative22\n",
    "with open('data/descriptives_list_positive.json', 'r') as f:\n",
    "    list_positive2 = json.load(f)\n",
    "\n",
    "# Filesname\n",
    "with open('data/filesname_negative.json', 'r') as f:\n",
    "    files_negatives11 = json.load(f)\n",
    "with open('data/filesname_drugs.json', 'r') as f:\n",
    "    files_negatives12 = json.load(f)\n",
    "files_negatives1 = files_negatives11 + files_negatives12\n",
    "with open('data/filesname_positive.json', 'r') as f:\n",
    "    files_positive = json.load(f)\n",
    "\n",
    "    \n",
    "num_negative = len(files_negatives1)\n",
    "num_positive = len(files_positive)\n",
    "\n",
    "list_negative = [ u + v for u,v in zip(list_negative1, list_negative2)]\n",
    "list_positive = [ u + v for u,v in zip(list_positive1, list_positive2)]\n",
    "\n",
    "noun_list = list_negative + list_positive\n",
    "filesname = files_negatives + files_positive\n",
    "\n",
    "# Building the mask, ie splitting among train and test set\n",
    "prop_train = 0.7\n",
    "\n",
    "# Negative\n",
    "mask_negative = np.zeros(num_negative, dtype=bool)\n",
    "mask_negative[:int(prop_train * num_negative)] = True\n",
    "np.random.shuffle(mask_negative)\n",
    "\n",
    "# Positive\n",
    "mask_positive = np.zeros(num_positive, dtype=bool)\n",
    "mask_positive[:int(prop_train * (num_positive))] = True\n",
    "np.random.shuffle(mask_positive)\n",
    "\n",
    "mask = np.concatenate((mask_negative, mask_positive))\n",
    "print(sum(mask))\n",
    "print(int(prop_train * (num_negative)) + int(prop_train * (num_positive)))\n",
    "\n",
    "# SVM model\n",
    "\n",
    "# Variables:\n",
    "# noun_list, num_topics, mask\n",
    "num_topics = 50\n",
    "\n",
    "# Building the output column\n",
    "Y = np.concatenate((np.zeros(num_negative), np.ones(num_positive)))\n",
    "\n",
    "# Baseline\n",
    "baseline_acc_train = accuracy_score(Y[np.where(mask)], np.zeros(sum(mask)), normalize=True, sample_weight=None)\n",
    "baseline_acc_test = accuracy_score(Y[np.where(~mask)], np.zeros(sum(~mask)), normalize=True, sample_weight=None)\n",
    "print('Baseline (All negative): score on train {}'.format(baseline_acc_train))\n",
    "print('Baseline (All negative): score on test {}'.format(baseline_acc_test))\n",
    "\n",
    "# SVM on Frequency matrix\n",
    "X_tf, count_vect2, tf_transformer2 = get_features(noun_list)\n",
    "\n",
    "svm_occ2 = SVC(kernel = 'linear', probability=True)\n",
    "svm_occ2.fit(X_tf[np.where(mask)], Y[np.where(mask)])\n",
    "print 'SVM occ: score on train {}'.format(svm_occ2.score(X_tf[np.where(mask)], Y[np.where(mask)]))\n",
    "print 'SVM occ: score on test {}'.format(svm_occ2.score(X_tf[np.where(~mask)], Y[np.where(~mask)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check that the drug tested are not in the test set\n",
    "for u,n in zip(urls, names):\n",
    "    X_c = get_counts(u, count_vect2)\n",
    "    print 'Prediction on {}: {}'.format(n, predict_occ(X_c, tf_transformer2, svm_occ2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Building a final classifier (called v2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Baseline (All negative): score on train 0.743680555556\n",
      "SVM occ: score on train 0.995972222222\n",
      "CPU times: user 1min 54s, sys: 1.14 s, total: 1min 55s\n",
      "Wall time: 1min 59s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "### Training a new model with negative training set = negative + drugs \n",
    "# with a test set (to validate the accuracy)\n",
    "\n",
    "# Loading the pre_processed training data\n",
    "# Nouns\n",
    "with open('data/noun_list_negative.json', 'r') as f:\n",
    "    list_negative11 = json.load(f)\n",
    "with open('data/noun_list_drugs.json', 'r') as f:\n",
    "    list_negative12 = json.load(f)\n",
    "list_negative1 = list_negative11 + list_negative12\n",
    "with open('data/noun_list_positive.json', 'r') as f:\n",
    "    list_positive1 = json.load(f)\n",
    "\n",
    "# Descriptives\n",
    "with open('data/descriptives_list_negative.json', 'r') as f:\n",
    "    list_negative21 = json.load(f)\n",
    "with open('data/descriptives_list_drugs.json', 'r') as f:\n",
    "    list_negative22 = json.load(f)\n",
    "list_negative2 = list_negative21 + list_negative22\n",
    "with open('data/descriptives_list_positive.json', 'r') as f:\n",
    "    list_positive2 = json.load(f)\n",
    "\n",
    "# Filesname\n",
    "with open('data/filesname_negative.json', 'r') as f:\n",
    "    files_negatives11 = json.load(f)\n",
    "with open('data/filesname_drugs.json', 'r') as f:\n",
    "    files_negatives12 = json.load(f)\n",
    "files_negatives1 = files_negatives11 + files_negatives12\n",
    "with open('data/filesname_positive.json', 'r') as f:\n",
    "    files_positive = json.load(f)\n",
    "\n",
    "    \n",
    "num_negative = len(files_negatives1)\n",
    "num_positive = len(files_positive)\n",
    "\n",
    "list_negative = [ u + v for u,v in zip(list_negative1, list_negative2)]\n",
    "list_positive = [ u + v for u,v in zip(list_positive1, list_positive2)]\n",
    "\n",
    "noun_list = list_negative + list_positive\n",
    "filesname = files_negatives + files_positive\n",
    "\n",
    "# SVM model\n",
    "\n",
    "# Variables:\n",
    "# noun_list, num_topics, mask\n",
    "num_topics = 50\n",
    "\n",
    "# Building the output column\n",
    "Y = np.concatenate((np.zeros(num_negative), np.ones(num_positive)))\n",
    "\n",
    "# Baseline\n",
    "baseline_acc_train = accuracy_score(Y, np.zeros(Y.shape[0]), normalize=True, sample_weight=None)\n",
    "print('Baseline (All negative): score on train {}'.format(baseline_acc_train))\n",
    "\n",
    "# SVM on Frequency matrix\n",
    "X_tf, count_vect2, tf_transformer2 = get_features(noun_list)\n",
    "\n",
    "svm_occ2 = SVC(kernel = 'linear', probability=True)\n",
    "svm_occ2.fit(X_tf, Y)\n",
    "print 'SVM occ: score on train {}'.format(svm_occ2.score(X_tf, Y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['model_v2/tf_transformer.pkl']"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Saving the pre-trained occurence svm model\n",
    "from sklearn.externals import joblib\n",
    "\n",
    "# Saving models\n",
    "joblib.dump(svm_occ2, 'model_v2/svm_occ.pkl')\n",
    "joblib.dump(count_vect2, 'model_v2/count_vect.pkl')\n",
    "joblib.dump(tf_transformer2, 'model_v2/tf_transformer.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prediction on new data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###########\n",
    "### Helper for prediction\n",
    "###########\n",
    "\n",
    "# Predict probability response of the occurrences model\n",
    "def predict_occ(X_counts, tf_transformer, svc):\n",
    "    # Feature Extraction\n",
    "    X_train_tf = tf_transformer.transform(X_counts)\n",
    "    # Prediction\n",
    "    y = svc.predict_proba(X_train_tf)[0][0]\n",
    "    return y\n",
    "\n",
    "# Predict probability response of the lda model\n",
    "def predict_lda(X_counts, lda_ge, svc_lda):\n",
    "    dt_new = get_lda_features(X_counts, lda_ge)\n",
    "    y1 = svc_lda.predict_proba(dt_new)[0][0]\n",
    "    \n",
    "    return y1\n",
    "\n",
    "# Retrieve the count matrix with the learned count_vect from the url\n",
    "def get_counts(url, count_vect):\n",
    "    t = requests.get(url)\n",
    "    paragraph, toc = get_text(t.text)\n",
    "    n, d, p = get_words(paragraph)\n",
    "    X_counts = count_vect.transform([n + d])\n",
    "    return X_counts\n",
    "\n",
    "# Prediction with the stacked regression on new data from url\n",
    "def predict_url_stacked(url, count_vect, tf_transformer, svc_occ, lda_ge, svc_lda, svc_stacked):\n",
    "    X_counts = get_counts(url, count_vect)\n",
    "    # Occurrence model\n",
    "    y_occ = predict_occ(X_counts, tf_transformer, svc_occ)\n",
    "    # Lda model\n",
    "    y_lda = predict_lda(X_counts, lda_ge, svc_lda)\n",
    "    # Stacked model\n",
    "    y = svc_stacked.predict(np.array([[y_occ, y_lda]]))\n",
    "    \n",
    "    return y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Edges cases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug missclassification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Defining the drugs\n",
    "urls = ['https://en.wikipedia.org/wiki/Penicillin', 'https://en.wikipedia.org/wiki/Paracetamol',\n",
    "        'https://en.wikipedia.org/wiki/L-DOPA', 'https://en.wikipedia.org/wiki/Erythromycin']\n",
    "names = [u.split('/')[-1] for u in urls]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prediction on Penicillin: 0.110800922733\n",
      "Prediction on Penicillin: 0.974746035987\n",
      "Prediction on Penicillin: [ 0.]\n",
      "Prediction on Paracetamol: 0.268033040857\n",
      "Prediction on Paracetamol: 0.949935650229\n",
      "Prediction on Paracetamol: [ 0.]\n",
      "Prediction on L-DOPA: 0.134786096126\n",
      "Prediction on L-DOPA: 0.968285484337\n",
      "Prediction on L-DOPA: [ 0.]\n",
      "Prediction on Erythromycin: 0.108924121663\n",
      "Prediction on Erythromycin: 0.995330926366\n",
      "Prediction on Erythromycin: [ 0.]\n"
     ]
    }
   ],
   "source": [
    "# Comparison of our 3 models on the drug classification task\n",
    "\n",
    "for u,n in zip(urls, names):\n",
    "    X_c = get_counts(u, count_vect)\n",
    "    print 'Prediction on {}: {}'.format(n, predict_lda(X_c, lda_ge, svm_lda))\n",
    "    print 'Prediction on {}: {}'.format(n, predict_occ(X_c, tf_transformer, svm_occ))\n",
    "    y = predict_url_stacked(u, count_vect, tf_transformer, svm_occ, lda_ge, svm_lda, svm_stacked)\n",
    "    print 'Prediction on {}: {}'.format(n, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generic class of disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 577,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open('noun_list_negative.json', 'r') as f:\n",
    "    nouns_list_negative = json.load(f)\n",
    "with open('noun_list_positive2.json', 'r') as f:\n",
    "    nouns_list_positive = json.load(f)\n",
    "\n",
    "with open('filesname_positive.json', 'r') as f:\n",
    "    files_positive = json.load(f)\n",
    "with open('filesname_negative.json', 'r') as f:\n",
    "    files_negatives = json.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 597,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "generic_words = ['virus', 'infection', 'bacteria', 'mutation', 'genetic', 'disorders']\n",
    "for i,f in enumerate(files_positive):\n",
    "    for w in generic_words:\n",
    "        if w in f:\n",
    "            files_positive.remove(f)\n",
    "            nouns_list_positive.remove(nouns_list_positive[i])\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "num_negative = len(files_negatives)\n",
    "num_positive = len(files_positive)\n",
    "\n",
    "noun_list = nouns_list_negative + nouns_list_positive\n",
    "filesname = files_negatives + files_positive\n",
    "\n",
    "# Building the output column\n",
    "Y_generic = np.concatenate((np.zeros(num_negative), np.ones(num_positive)))\n",
    "\n",
    "X_tf_generic, count_vect_generic, tf_transformer_generic = get_features(noun_list, filesname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9552\n",
      "9552\n"
     ]
    }
   ],
   "source": [
    "# Splitting among train and test set\n",
    "# First choice of 70 - 30\n",
    "prop_train = 0.7\n",
    "\n",
    "# Negative\n",
    "mask_negative = np.zeros(num_negative, dtype=bool)\n",
    "mask_negative[:int(prop_train * num_negative)] = True\n",
    "np.random.shuffle(mask_negative)\n",
    "\n",
    "# Positive\n",
    "mask_positive = np.zeros(num_positive, dtype=bool)\n",
    "mask_positive[:int(prop_train * (num_positive))] = True\n",
    "np.random.shuffle(mask_positive)\n",
    "\n",
    "mask = np.concatenate((mask_negative, mask_positive))\n",
    "print(sum(mask))\n",
    "print(int(prop_train * (num_negative + num_positive)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Prediction with broad articles ', True)\n",
      "('Prediction without broad articles ', True)\n"
     ]
    }
   ],
   "source": [
    "# Comparaison of the full classifier and the one with page removed\n",
    "y_generic = predict_url('https://en.wikipedia.org/wiki/Cancer', count_vect_generic, tf_transformer_generic, svc_generic)\n",
    "y = predict_url('https://en.wikipedia.org/wiki/Cancer', count_vect, tf_transformer, svc)\n",
    "print('Prediction with broad articles ', y)\n",
    "print('Prediction without broad articles ', y_generic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Building a dataframe with the article name (not storing the features because in sparse format)\n",
    "\n",
    "# Building the prediction vector\n",
    "\n",
    "Y_pred = np.zeros(len(Y))\n",
    "Y_pred[mask] = svm_stacked.predict(X_lr[np.where(mask)])\n",
    "Y_pred[~mask] = svm_stacked.predict(X_lr[np.where(~mask)])\n",
    "\n",
    "# Processing the titles\n",
    "with open('title_listnegative.json', 'r') as f:\n",
    "    titles_negatives = json.load(f)\n",
    "with open('title_listpositive.json', 'r') as f:\n",
    "    titles_positive = json.load(f)\n",
    "\n",
    "titles = titles_negatives + titles_positive\n",
    "\n",
    "df = pd.DataFrame(data={'True': Y, 'Prediction': Y_pred, 'Train':mask, 'Filename':filesname, 'Titles':titles})\n",
    "df.Filename = df.Filename.astype(unicode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Filename</th>\n",
       "      <th>Prediction</th>\n",
       "      <th>Titles</th>\n",
       "      <th>Train</th>\n",
       "      <th>True</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>1152</td>\n",
       "      <td>1</td>\n",
       "      <td>Rapid eye movement behavior disorder</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>135</td>\n",
       "      <td>1</td>\n",
       "      <td>LIN28</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>1616</td>\n",
       "      <td>1</td>\n",
       "      <td>Pre-B-cell leukemia homeobox</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>889</th>\n",
       "      <td>1799</td>\n",
       "      <td>1</td>\n",
       "      <td>Geriatric trauma</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1094</th>\n",
       "      <td>1983</td>\n",
       "      <td>1</td>\n",
       "      <td>Occult pneumonia</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>2017</td>\n",
       "      <td>1</td>\n",
       "      <td>Tranilast</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880</th>\n",
       "      <td>2690</td>\n",
       "      <td>1</td>\n",
       "      <td>Osteomyelitis of the jaws</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895</th>\n",
       "      <td>2703</td>\n",
       "      <td>1</td>\n",
       "      <td>List of aquarium diseases</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2305</th>\n",
       "      <td>3072</td>\n",
       "      <td>1</td>\n",
       "      <td>Protozoan infection</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2631</th>\n",
       "      <td>3366</td>\n",
       "      <td>1</td>\n",
       "      <td>Aphonia</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2838</th>\n",
       "      <td>3552</td>\n",
       "      <td>1</td>\n",
       "      <td>Charles R. Attwood</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3032</th>\n",
       "      <td>3727</td>\n",
       "      <td>1</td>\n",
       "      <td>Pectus carinatum</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3240</th>\n",
       "      <td>3914</td>\n",
       "      <td>1</td>\n",
       "      <td>Arcanobacterium haemolyticum</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>4309</td>\n",
       "      <td>1</td>\n",
       "      <td>Radotinib</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4977</th>\n",
       "      <td>5478</td>\n",
       "      <td>1</td>\n",
       "      <td>Retinal waves</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5011</th>\n",
       "      <td>5508</td>\n",
       "      <td>1</td>\n",
       "      <td>Micromonospora fulviviridis</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5389</th>\n",
       "      <td>5849</td>\n",
       "      <td>1</td>\n",
       "      <td>Treponema carateum</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5518</th>\n",
       "      <td>5965</td>\n",
       "      <td>1</td>\n",
       "      <td>Jod-Basedow phenomenon</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5674</th>\n",
       "      <td>6104</td>\n",
       "      <td>1</td>\n",
       "      <td>Apocrine nevus</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5818</th>\n",
       "      <td>6234</td>\n",
       "      <td>1</td>\n",
       "      <td>Mad Travelers: Reflections on the Reality of T...</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6098</th>\n",
       "      <td>6487</td>\n",
       "      <td>1</td>\n",
       "      <td>SOCRATES (pain assessment)</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6201</th>\n",
       "      <td>658</td>\n",
       "      <td>1</td>\n",
       "      <td>Graves' ophthalmopathy</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6858</th>\n",
       "      <td>7170</td>\n",
       "      <td>1</td>\n",
       "      <td>Mammary myofibroblastoma</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6890</th>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>Transforming growth interacting factor</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7212</th>\n",
       "      <td>749</td>\n",
       "      <td>1</td>\n",
       "      <td>Microshock</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7712</th>\n",
       "      <td>794</td>\n",
       "      <td>1</td>\n",
       "      <td>Heyde's syndrome</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7958</th>\n",
       "      <td>8160</td>\n",
       "      <td>1</td>\n",
       "      <td>Henneke horse body condition scoring system</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8482</th>\n",
       "      <td>8632</td>\n",
       "      <td>1</td>\n",
       "      <td>Online patient education</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8641</th>\n",
       "      <td>8776</td>\n",
       "      <td>1</td>\n",
       "      <td>Antibiotic candidiasis</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8897</th>\n",
       "      <td>9005</td>\n",
       "      <td>1</td>\n",
       "      <td>Polychromia</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11180</th>\n",
       "      <td>Epiphyseal_stippling_syndrome_osteoclastic_hyp...</td>\n",
       "      <td>0</td>\n",
       "      <td>Pacman dysplasia</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11335</th>\n",
       "      <td>Fluorosis</td>\n",
       "      <td>0</td>\n",
       "      <td>Fluorosis</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11371</th>\n",
       "      <td>Fructose-1,6-bisphosphatase_deficiency</td>\n",
       "      <td>0</td>\n",
       "      <td>Fructose 1,6-bisphosphatase</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11703</th>\n",
       "      <td>Hutchinson_incisors</td>\n",
       "      <td>0</td>\n",
       "      <td>Hutchinson's teeth</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11827</th>\n",
       "      <td>Ichthyoallyeinotoxism</td>\n",
       "      <td>0</td>\n",
       "      <td>Ichthyoallyeinotoxism</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11869</th>\n",
       "      <td>Infantile_digital_fibromatosis</td>\n",
       "      <td>0</td>\n",
       "      <td>Infantile digital fibromatosis</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12013</th>\n",
       "      <td>Launois%E2%80%93Bensaude_adenolipomatosis</td>\n",
       "      <td>0</td>\n",
       "      <td>Raoul Bensaude</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12206</th>\n",
       "      <td>Megaduodenum</td>\n",
       "      <td>0</td>\n",
       "      <td>Megaduodenum</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12226</th>\n",
       "      <td>Mesomelia</td>\n",
       "      <td>0</td>\n",
       "      <td>Mesomelia</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12362</th>\n",
       "      <td>Mycetoma</td>\n",
       "      <td>0</td>\n",
       "      <td>Mycetoma</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12398</th>\n",
       "      <td>Myxoid_liposarcoma</td>\n",
       "      <td>0</td>\n",
       "      <td>Myxoid liposarcoma</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12464</th>\n",
       "      <td>Nicotine_withdrawal</td>\n",
       "      <td>0</td>\n",
       "      <td>Nicotine withdrawal</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12489</th>\n",
       "      <td>Nut_allergy</td>\n",
       "      <td>0</td>\n",
       "      <td>Tree nut allergy</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12509</th>\n",
       "      <td>Ocular_albinism</td>\n",
       "      <td>0</td>\n",
       "      <td>Ocular albinism</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12610</th>\n",
       "      <td>Pacman_dysplasia</td>\n",
       "      <td>0</td>\n",
       "      <td>Pacman dysplasia</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12738</th>\n",
       "      <td>Phosphoglucomutase_deficiency</td>\n",
       "      <td>0</td>\n",
       "      <td>Phosphoglucomutase</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12742</th>\n",
       "      <td>Physical_urticaria</td>\n",
       "      <td>0</td>\n",
       "      <td>Physical urticaria</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12762</th>\n",
       "      <td>PIRA</td>\n",
       "      <td>0</td>\n",
       "      <td>Provisional Irish Republican Army</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12783</th>\n",
       "      <td>Pneumonoultramicroscopicsilicovolcanoconiosis</td>\n",
       "      <td>0</td>\n",
       "      <td>Pneumonoultramicroscopicsilicovolcanoconiosis</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12840</th>\n",
       "      <td>Posterior_tibial_tendon_rupture</td>\n",
       "      <td>0</td>\n",
       "      <td>Tibialis posterior muscle</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12955</th>\n",
       "      <td>Pyrosis</td>\n",
       "      <td>0</td>\n",
       "      <td>Pyrosis</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13029</th>\n",
       "      <td>Rhabditida_infections</td>\n",
       "      <td>0</td>\n",
       "      <td>Rhabditida</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13044</th>\n",
       "      <td>Rickettsiosis</td>\n",
       "      <td>0</td>\n",
       "      <td>Rickettsiosis</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13141</th>\n",
       "      <td>Secernentea_Infections</td>\n",
       "      <td>0</td>\n",
       "      <td>Secernentea</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13190</th>\n",
       "      <td>Shwartzman_phenomenon</td>\n",
       "      <td>0</td>\n",
       "      <td>Shwartzman phenomenon</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13192</th>\n",
       "      <td>Sialadenitis</td>\n",
       "      <td>0</td>\n",
       "      <td>Sialadenitis</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13198</th>\n",
       "      <td>Silent_sinus_syndrome</td>\n",
       "      <td>0</td>\n",
       "      <td>Silent sinus syndrome</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13228</th>\n",
       "      <td>Somatostatinoma</td>\n",
       "      <td>0</td>\n",
       "      <td>Somatostatinoma</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13401</th>\n",
       "      <td>Togaviridae_disease</td>\n",
       "      <td>0</td>\n",
       "      <td>Togaviridae</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13519</th>\n",
       "      <td>Uveal_diseases</td>\n",
       "      <td>0</td>\n",
       "      <td>Uvea</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Filename  Prediction  \\\n",
       "172                                                 1152           1   \n",
       "391                                                  135           1   \n",
       "687                                                 1616           1   \n",
       "889                                                 1799           1   \n",
       "1094                                                1983           1   \n",
       "1133                                                2017           1   \n",
       "1880                                                2690           1   \n",
       "1895                                                2703           1   \n",
       "2305                                                3072           1   \n",
       "2631                                                3366           1   \n",
       "2838                                                3552           1   \n",
       "3032                                                3727           1   \n",
       "3240                                                3914           1   \n",
       "3679                                                4309           1   \n",
       "4977                                                5478           1   \n",
       "5011                                                5508           1   \n",
       "5389                                                5849           1   \n",
       "5518                                                5965           1   \n",
       "5674                                                6104           1   \n",
       "5818                                                6234           1   \n",
       "6098                                                6487           1   \n",
       "6201                                                 658           1   \n",
       "6858                                                7170           1   \n",
       "6890                                                  72           1   \n",
       "7212                                                 749           1   \n",
       "7712                                                 794           1   \n",
       "7958                                                8160           1   \n",
       "8482                                                8632           1   \n",
       "8641                                                8776           1   \n",
       "8897                                                9005           1   \n",
       "...                                                  ...         ...   \n",
       "11180  Epiphyseal_stippling_syndrome_osteoclastic_hyp...           0   \n",
       "11335                                          Fluorosis           0   \n",
       "11371             Fructose-1,6-bisphosphatase_deficiency           0   \n",
       "11703                                Hutchinson_incisors           0   \n",
       "11827                              Ichthyoallyeinotoxism           0   \n",
       "11869                     Infantile_digital_fibromatosis           0   \n",
       "12013          Launois%E2%80%93Bensaude_adenolipomatosis           0   \n",
       "12206                                       Megaduodenum           0   \n",
       "12226                                          Mesomelia           0   \n",
       "12362                                           Mycetoma           0   \n",
       "12398                                 Myxoid_liposarcoma           0   \n",
       "12464                                Nicotine_withdrawal           0   \n",
       "12489                                        Nut_allergy           0   \n",
       "12509                                    Ocular_albinism           0   \n",
       "12610                                   Pacman_dysplasia           0   \n",
       "12738                      Phosphoglucomutase_deficiency           0   \n",
       "12742                                 Physical_urticaria           0   \n",
       "12762                                               PIRA           0   \n",
       "12783      Pneumonoultramicroscopicsilicovolcanoconiosis           0   \n",
       "12840                    Posterior_tibial_tendon_rupture           0   \n",
       "12955                                            Pyrosis           0   \n",
       "13029                              Rhabditida_infections           0   \n",
       "13044                                      Rickettsiosis           0   \n",
       "13141                             Secernentea_Infections           0   \n",
       "13190                              Shwartzman_phenomenon           0   \n",
       "13192                                       Sialadenitis           0   \n",
       "13198                              Silent_sinus_syndrome           0   \n",
       "13228                                    Somatostatinoma           0   \n",
       "13401                                Togaviridae_disease           0   \n",
       "13519                                     Uveal_diseases           0   \n",
       "\n",
       "                                                  Titles  Train  True  \n",
       "172                 Rapid eye movement behavior disorder  False     0  \n",
       "391                                                LIN28   True     0  \n",
       "687                         Pre-B-cell leukemia homeobox   True     0  \n",
       "889                                     Geriatric trauma   True     0  \n",
       "1094                                    Occult pneumonia  False     0  \n",
       "1133                                           Tranilast  False     0  \n",
       "1880                           Osteomyelitis of the jaws   True     0  \n",
       "1895                           List of aquarium diseases   True     0  \n",
       "2305                                 Protozoan infection   True     0  \n",
       "2631                                             Aphonia   True     0  \n",
       "2838                                  Charles R. Attwood  False     0  \n",
       "3032                                    Pectus carinatum   True     0  \n",
       "3240                        Arcanobacterium haemolyticum  False     0  \n",
       "3679                                           Radotinib  False     0  \n",
       "4977                                       Retinal waves  False     0  \n",
       "5011                         Micromonospora fulviviridis  False     0  \n",
       "5389                                  Treponema carateum   True     0  \n",
       "5518                              Jod-Basedow phenomenon   True     0  \n",
       "5674                                      Apocrine nevus  False     0  \n",
       "5818   Mad Travelers: Reflections on the Reality of T...  False     0  \n",
       "6098                          SOCRATES (pain assessment)  False     0  \n",
       "6201                              Graves' ophthalmopathy   True     0  \n",
       "6858                            Mammary myofibroblastoma   True     0  \n",
       "6890              Transforming growth interacting factor  False     0  \n",
       "7212                                          Microshock   True     0  \n",
       "7712                                    Heyde's syndrome  False     0  \n",
       "7958         Henneke horse body condition scoring system  False     0  \n",
       "8482                            Online patient education  False     0  \n",
       "8641                              Antibiotic candidiasis  False     0  \n",
       "8897                                         Polychromia  False     0  \n",
       "...                                                  ...    ...   ...  \n",
       "11180                                   Pacman dysplasia  False     1  \n",
       "11335                                          Fluorosis  False     1  \n",
       "11371                        Fructose 1,6-bisphosphatase   True     1  \n",
       "11703                                 Hutchinson's teeth  False     1  \n",
       "11827                              Ichthyoallyeinotoxism   True     1  \n",
       "11869                     Infantile digital fibromatosis  False     1  \n",
       "12013                                     Raoul Bensaude   True     1  \n",
       "12206                                       Megaduodenum  False     1  \n",
       "12226                                          Mesomelia  False     1  \n",
       "12362                                           Mycetoma   True     1  \n",
       "12398                                 Myxoid liposarcoma  False     1  \n",
       "12464                                Nicotine withdrawal  False     1  \n",
       "12489                                   Tree nut allergy  False     1  \n",
       "12509                                    Ocular albinism  False     1  \n",
       "12610                                   Pacman dysplasia  False     1  \n",
       "12738                                 Phosphoglucomutase  False     1  \n",
       "12742                                 Physical urticaria  False     1  \n",
       "12762                  Provisional Irish Republican Army   True     1  \n",
       "12783      Pneumonoultramicroscopicsilicovolcanoconiosis   True     1  \n",
       "12840                          Tibialis posterior muscle  False     1  \n",
       "12955                                            Pyrosis  False     1  \n",
       "13029                                         Rhabditida  False     1  \n",
       "13044                                      Rickettsiosis  False     1  \n",
       "13141                                        Secernentea   True     1  \n",
       "13190                              Shwartzman phenomenon  False     1  \n",
       "13192                                       Sialadenitis  False     1  \n",
       "13198                              Silent sinus syndrome  False     1  \n",
       "13228                                    Somatostatinoma  False     1  \n",
       "13401                                        Togaviridae   True     1  \n",
       "13519                                               Uvea  False     1  \n",
       "\n",
       "[73 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Missclassification: This is to try to understand the pages we missclassified.\n",
    "# It seems that many negative ones look like desease one so the confusion is not surprising.\n",
    "df[(df.Prediction != df.True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###########\n",
    "### Helper for information extraction\n",
    "###########\n",
    "\n",
    "\n",
    "def get_description_from_file(filename, path, kw_list):\n",
    "    # identify if the class is negative or positive\n",
    "    only_digit = re.compile('^[0-9]*$')\n",
    "    if bool(only_digit.search(filename)):\n",
    "        class_ = 'negative'\n",
    "    else:\n",
    "        class_ = 'positive'\n",
    "\n",
    "    with open(path + class_ + '/' + filename) as f:\n",
    "        data = f.read()\n",
    "    return get_information(t.text, kw_list)\n",
    "\n",
    "\n",
    "def get_description_from_url(url, kw_list):\n",
    "    t = requests.get(url)\n",
    "    title = get_title(t.text)\n",
    "    return title, get_information(t.text, kw_list)\n",
    "   \n",
    "\n",
    "# Return sentences containing specific key-words\n",
    "def get_information(data, kw_list):\n",
    "    # Extracting Content\n",
    "    paragraph, toc = get_text(data)\n",
    "    # Result\n",
    "    kw_to_info = {kw: u'' for kw in kw_list}\n",
    "\n",
    "    # Removing the possible link\n",
    "    sp_clean = [t for t in paragraph.split() if t not in ['[', ']'] and not contains_digits(t)]\n",
    "    paragraph_clean = ' '.join(sp_clean)\n",
    "\n",
    "    sentences = paragraph_clean.split('. ')\n",
    "    for s in sentences:\n",
    "        # Adding the content to corresponding kw if present\n",
    "        # (by default the first one is chosen)\n",
    "        for kw in kw_list:\n",
    "            if kw in s:\n",
    "                kw_to_info[kw] += s + u'. '\n",
    "                break\n",
    "    return kw_to_info\n",
    "\n",
    "\n",
    "# Final function to classify a new entry given a pre-trained model and extract its information\n",
    "def classify_from_url(url, count_vect, tf_transformer, svm_occ, kw_list):\n",
    "    # Classification\n",
    "    X_counts = get_counts(url, count_vect)\n",
    "    y = predict_occ(X_counts, tf_transformer, svm_occ)\n",
    "    res = bool(y<0.5)\n",
    "    # Extracting information if needed\n",
    "    if res:\n",
    "        title, description = get_description_from_url(url, kw_list)\n",
    "        return res, title, description\n",
    "    else:\n",
    "        return res, None, None\n",
    "\n",
    "\n",
    "# Print in a pretty way the dictionary of disease description\n",
    "def pretty_print_description(description):\n",
    "    print u'Description of the disease: \\n'\n",
    "    for k, v in description.iteritems():\n",
    "        if v:\n",
    "            print(k.upper())\n",
    "            print v + u' \\n'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/nicolasdrizard/anaconda/envs/python2/lib/python2.7/site-packages/ipykernel/__main__.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/nicolasdrizard/anaconda/envs/python2/lib/python2.7/site-packages/ipykernel/__main__.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 38.3 s, sys: 1.38 s, total: 39.7 s\n",
      "Wall time: 44.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Script to extract information for the prediction on the whole training set (for the entries predicted as a disease)\n",
    "kw_list = ['symptom', 'cause', 'prognosis', 'prevention', 'treatment', 'drug', 'susceptibility', 'feature', 'frequent']\n",
    "path = 'training/'\n",
    "\n",
    "disease_filename = df[(df['Prediction'] == 1) & df['True']].Filename\n",
    "\n",
    "# Extract disease name\n",
    "df['disease'] = ''\n",
    "df['disease'][(df.Prediction == True) | (df.True == True)] = df['Titles'][(df.Prediction == True) | (df.True == True)]\n",
    "\n",
    "# Extract disease description\n",
    "df['description'] = u''\n",
    "df['description'][(df.Prediction == True) | (df.True == True)] = (df[(df.Prediction == True) | (df.True == True)]).apply(lambda row: get_description_from_file(row.Filename, path, kw_list), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Filename</th>\n",
       "      <th>Prediction</th>\n",
       "      <th>Titles</th>\n",
       "      <th>Train</th>\n",
       "      <th>True</th>\n",
       "      <th>disease</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10000</th>\n",
       "      <td>Aagenaes_syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aagenaes syndrome</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>Aagenaes syndrome</td>\n",
       "      <td>Aagenaes syndrome is a syndrome characterised...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10001</th>\n",
       "      <td>Aarskog%E2%80%93Ose%E2%80%93Pande_syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>SHORT syndrome</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>SHORT syndrome</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10002</th>\n",
       "      <td>Aarskog_syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>AarskogScott syndrome</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>AarskogScott syndrome</td>\n",
       "      <td>The syndrome is caused by mutation in a gene c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10003</th>\n",
       "      <td>Aase%E2%80%93Smith_syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aase syndrome</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Aase syndrome</td>\n",
       "      <td>The anemia is caused by underdevelopment of th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10004</th>\n",
       "      <td>Aase_syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aase syndrome</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Aase syndrome</td>\n",
       "      <td>The anemia is caused by underdevelopment of th...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Filename  Prediction  \\\n",
       "10000                           Aagenaes_syndrome           1   \n",
       "10001  Aarskog%E2%80%93Ose%E2%80%93Pande_syndrome           1   \n",
       "10002                            Aarskog_syndrome           1   \n",
       "10003                 Aase%E2%80%93Smith_syndrome           1   \n",
       "10004                               Aase_syndrome           1   \n",
       "\n",
       "                       Titles  Train  True                 disease  \\\n",
       "10000       Aagenaes syndrome  False     1       Aagenaes syndrome   \n",
       "10001          SHORT syndrome   True     1          SHORT syndrome   \n",
       "10002  AarskogScott syndrome   True     1  AarskogScott syndrome   \n",
       "10003           Aase syndrome   True     1           Aase syndrome   \n",
       "10004           Aase syndrome   True     1           Aase syndrome   \n",
       "\n",
       "                                             description  \n",
       "10000   Aagenaes syndrome is a syndrome characterised...  \n",
       "10001                                                     \n",
       "10002  The syndrome is caused by mutation in a gene c...  \n",
       "10003  The anemia is caused by underdevelopment of th...  \n",
       "10004  The anemia is caused by underdevelopment of th...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Illustration on 5 well classified entries\n",
    "df[(df.Prediction == True) & (df.True == True)].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prediction on Paracetamol: Non Disease \n",
      "\n",
      "Prediction on Aagenaes_syndrome: Disease \n",
      "\n",
      "Name of the disease: Aagenaes syndrome\n",
      "\n",
      "Description of the disease: \n",
      "\n",
      "CAUSE\n",
      "Aagenaes syndrome is a syndrome characterised by congenital hypoplasia of lymph vessels , which causes lymphedema of the legs and recurrent cholestasis in infancy, and slow progress to hepatic cirrhosis and giant cell hepatitis with fibrosis of the portal tracts . The genetic cause is unknown, but it is autosomal recessively inherited and the gene is located to chromosome .  \n",
      "\n",
      "Prediction on Cancer: Disease \n",
      "\n",
      "Name of the disease: Cancer\n",
      "\n",
      "Description of the disease: \n",
      "\n",
      "DRUG\n",
      "The term encompasses any of a large variety of different anticancer drugs, which are divided into broad categories such as alkylating agents and antimetabolites . However, radiation and radioactive drugs are normally avoided during pregnancy, especially if the fetal dose might exceed cGy. Also, when chemotherapy is being given after birth, many of the drugs pass through breast milk to the baby, which could harm the baby.  \n",
      "\n",
      "PROGNOSIS\n",
      "Together, this information is useful to evaluate the prognosis of the patient and to choose the best treatment. These tests may provide information about the molecular changes (such as mutations , fusion genes , and numerical chromosome changes) that have happened in the cancer cells, and may thus also indicate the future behavior of the cancer (prognosis) and best treatment. Once a cancer has metastasized or spread beyond its original site, the prognosis normally becomes much worse. This view of the disease was first formulated by the English surgeon Campbell De Morgan between and Though many diseases (such as heart failure ) may have a worse prognosis than most cases of cancer, cancer is the subject of widespread fear and taboos.  \n",
      "\n",
      "SYMPTOM\n",
      "Possible signs and symptoms include: a new lump, abnormal bleeding, a prolonged cough, unexplained weight loss , and a change in bowel movements among others. While these symptoms may indicate cancer, they may also occur due to other issues. Cancer can be detected by certain signs and symptoms or screening tests . Pain and symptom management are an important part of care. When cancer begins, it invariably produces no symptoms. Signs and symptoms only appear as the mass continues to grow or ulcerates . Few symptoms are specific , with many of them also frequently occurring in individuals who have other conditions. Thus, it is not uncommon for people diagnosed with cancer to have been treated for other diseases, which were assumed to be causing their symptoms. Local symptoms may occur due to the mass of the tumor or its ulceration. General symptoms occur due to distant effects of the cancer that are not related to direct or metastatic spread. Some cancers may cause specific groups of systemic symptoms, termed paraneoplastic phenomena . The symptoms of metastatic cancers depend on the location of the tumor, and can include enlarged lymph nodes (which can be felt or sometimes seen under the skin and are typically hard), enlarged liver or enlarged spleen , which can be felt in the abdomen , pain or fracture of affected bones, and neurological symptoms. Most cancers are initially recognized either because of the appearance of signs or symptoms or through screening . Unlike diagnosis efforts prompted by symptoms and medical signs , cancer screening involves efforts to detect cancer after it has formed, but before any noticeable symptoms appear. Even when it is impossible for chemotherapy to provide a permanent cure, chemotherapy may be useful to reduce symptoms like pain or to reduce the size of an inoperable tumor in the hope that surgery will be possible in the future. Radiation therapy involves the use of ionizing radiation in an attempt to either cure or improve the symptoms of cancer. Multiple national medical guidelines recommend early palliative care for people whose cancer has produced distressing symptoms (pain, shortness of breath, fatigue, nausea) or who need help coping with their illness. Diagnosing a new cancer in a pregnant woman is difficult, in part because any symptoms are commonly assumed to be a normal discomfort associated with pregnancy.  \n",
      "\n",
      "DIAGNOSIS\n",
      "In children under at diagnosis the five-year survival rate in the developed world is on average For cancer in the United States the average five-year survival rate is In about million new cases of cancer occurred globally (not including skin cancer other than melanoma ). Neither of these lead to a definitive diagnosis, which requires the examination of a tissue sample by a pathologist . The tissue diagnosis given by the pathologist indicates the type of cell that is proliferating, its histological grade , genetic abnormalities, and other features of the tumor. It is typically an important part of making the definitive diagnosis and staging the tumor as biopsies are usually required. However, the survival rates vary dramatically by type of cancer, and by the stage at which it is diagnosed, with the range running from the great majority of people surviving to almost no one surviving as long as five years after diagnosis.  \n",
      "\n",
      "CAUSE\n",
      "Tobacco use is the cause of about of cancer deaths. It caused about million deaths or of all human deaths . For example, mass effects from lung cancer can cause blockage of the bronchus resulting in cough or pneumonia ; esophageal cancer can cause narrowing of the esophagus, making it difficult or painful to swallow; and colorectal cancer may lead to narrowing or blockages in the bowel, resulting in changes in bowel habits. Ulceration can cause bleeding that, if it occurs in the lung, will lead to coughing up blood , in the bowels to anemia or rectal bleeding , in the bladder to blood in the urine , and in the uterus to vaginal bleeding. Some cancers can cause a buildup of fluid within the chest or abdomen . Hodgkin disease , leukemias , and cancers of the liver or kidney can cause a persistent fever of unknown origin . Environmental , as used by cancer researchers, means any cause that is not inherited genetically , such as lifestyle, economic and behavioral factors, and not merely pollution. It is nearly impossible to prove what caused a cancer in any individual, because most cancers have multiple possible causes. For example, if a person who uses tobacco heavily develops lung cancer, then it was probably caused by the tobacco use, but since everyone has a small chance of developing lung cancer as a result of air pollution or radiation, then there is a small chance that the cancer developed because of air pollution or radiation. Tobacco smoking , for example, causes of lung cancer . It also causes cancer in the larynx , head, neck, stomach, bladder, kidney, esophagus and pancreas . Aflatoxin , a frequent food contaminate, causes liver cancer. Betel nut chewing causes oral cancer. Viruses are the usual infectious agents that cause cancer but cancer bacteria and parasites may also have an effect. A virus that can cause cancer is called an oncovirus . Additionally, the vast majority of non-invasive cancers are non-melanoma skin cancers caused by non-ionizing ultraviolet radiation , mostly from sunlight. Radiation can cause cancer in most parts of the body, in all animals, and at any age. Clear evidence establishes ultraviolet radiation, especially the non-ionizing medium wave UVB , as the cause of most non-melanoma skin cancers , which are the most common forms of cancer in the world. Hereditary cancers are primarily caused by an inherited genetic defect. Less than of the population are carriers of a genetic mutation that has a large effect on cancer risk and these cause less than of all cancer. Some substances cause cancer primarily through their physical, rather than chemical, effects on cells. A prominent example of this is prolonged exposure to asbestos , naturally occurring mineral fibers that are a major cause of mesothelioma , which is a cancer of the serous membrane , usually the serous membrane surrounding the lungs. Non-fibrous particulate materials that cause cancer include powdered metallic cobalt and nickel , and crystalline silica ( quartz , cristobalite , and tridymite ). Similarly, physical trauma is not accepted as a cause for cervical cancer, breast cancer, or brain cancer. Generally, it is believed that the cancer arises, or a pre-existing cancer is encouraged, during the process of repairing the trauma, rather than the cancer being caused directly by the trauma. It is controversial whether chronic inflammation can directly cause mutation. Some treatments and prevention approaches leverage this cause by artificially reducing hormone levels, and thus discouraging hormone-sensitive cancers. The errors that cause cancer are self-amplifying and compounding , for example: The transformation of normal cell into cancer is akin to a chain reaction caused by initial errors, which compound into more severe errors, each progressively allowing the cell to escape the controls that limit normal tissue growth. Examples of such modifications are changes in DNA methylation (hypermethylation and hypomethylation) and histone modification and changes in chromosomal architecture (caused by inappropriate expression of proteins such as or ). Such alterations are thought to occur early in progression to cancer and to be a likely cause of the genetic instability characteristic of cancers. Reduced expression of DNA repair genes causes deficient DNA repair. (In the figure, red wording indicates the central role of DNA damage and defects in DNA repair in progression to cancer.) When DNA repair is deficient DNA damages remain in cells at a higher than usual level level from the top in figure), and these excess damages cause increased frequencies of mutation and/or epimutation level from top of figure). Higher levels of DNA damage not only cause increased mutation (right side of figure), but also cause increased epimutation. During repair of DNA double strand breaks, or repair of other DNA damages, incompletely cleared sites of repair can cause epigenetic gene silencing. Deficient expression of DNA repair proteins due to an inherited mutation can cause an increased risk of cancer. However, such germline mutations (which cause highly penetrant cancer syndromes) are the cause of only about percent of cancers. In sporadic cancers, deficiencies in DNA repair are occasionally caused by a mutation in a DNA repair gene, but are much more frequently caused by epigenetic alterations that reduce or silence expression of DNA repair genes. Many studies of heavy metal-induced carcinogenesis show that such heavy metals cause reduction in expression of DNA repair enzymes, some through epigenetic mechanisms. As pointed out above under genetic alterations, cancer is caused by failure to regulate tissue growth, when the genes that regulate cell growth and differentiation are altered. It has become clear that these alterations are caused by both DNA sequence mutation in oncogenes and tumor suppressor genes as well as by epigenetic alterations. The epigenetic deficiencies in expression of DNA repair genes, in particular, likely cause an increased frequency of mutations, some of which then occur in oncogenes and tumor suppressor genes. Not all environmental causes are controllable, such as naturally occurring background radiation , and other cases of cancer are caused through hereditary genetic disorders, and thus it is not possible to prevent all cases of cancer. In the general population, NSAIDs reduce the risk of colorectal cancer , however due to the cardiovascular and gastrointestinal side effects they cause overall harm when used for prevention. Survival is worse in the developing world, partly because the types of cancer that are most common there are at present harder to treat than those associated with the lifestyle of developed countries. Additionally, patients with worse prognoses may be depressed or report a lower quality of life directly because they correctly perceive that their condition is likely to be fatal. This makes invasive cancer the leading cause of death in the developed world and the second leading cause of death in the developing world . Autopsy studies in Europe and Asia have shown that up to of people have undiagnosed and apparently harmless thyroid cancer at the time of their deaths, and that of men develop prostate cancer by age As these cancers did not cause the person's death, identifying them would have represented overdiagnosis rather than useful medical care. Galen stated that \"cancer of the breast is so called because of the fancied resemblance to a crab given by the lateral prolongations of the tumor and the adjacent distended veins\". In the and centuries, it became acceptable for doctors to dissect bodies to discover the cause of death. The German professor Wilhelm Fabry believed that breast cancer was caused by a milk clot in a mammary duct. The Dutch professor Francois de la Boe Sylvius , a follower of Descartes , believed that all disease was the outcome of chemical processes, and that acidic lymph fluid was the cause of cancer. The physician John Hill described tobacco snuff as the cause of nose cancer in This was followed by the report in by British surgeon Percivall Pott that chimney sweeps' carcinoma , a cancer of the scrotum , was a common disease among chimney sweeps . This deep belief that cancer is necessarily a difficult and usually deadly disease is reflected in the systems chosen by society to compile cancer statistics: the most common form of cancernon-melanoma skin cancers , accounting for about one-third of all cancer cases worldwide, but very few deaths are excluded from cancer statistics specifically because they are easily treated and almost always cured, often in a single, short, outpatient procedure. In the a relatively popular alternative cancer treatment was a specialized form of talk therapy , based on the idea that cancer was caused by a bad attitude. Among other effects, this belief allows society to blame the victim for having caused the cancer (by \"wanting\" it) or having prevented its cure (by not becoming a sufficiently happy, fearless, and loving person). This is based upon the blamers' belief that the world is fundamentally just, and so any dangerous illness, like cancer, must be a type of punishment for bad choices, because in a just world, bad things would not happen to good people. In of Hispanics and of children years old or younger lacked health insurance; \"uninsured patients and those from ethnic minorities are substantially more likely to be diagnosed with cancer at a later stage, when treatment can be more extensive and more costly.\" Because cancer is a class of diseases, it is unlikely that there will ever be a single \" cure for cancer \" any more than there will be a single treatment for all infectious diseases . Research about cancer causes focuses on the following issues: The improved understanding of molecular biology and cellular biology due to cancer research has led to a number of new treatments for cancer since U.S. Because cancer is largely a disease of older adults, it is not common in pregnant women. Surgery is generally safe, but pelvic surgeries during the first trimester may cause miscarriage.  \n",
      "\n",
      "PREVENTION\n",
      "Complex error correction and prevention is built into the process, and safeguards the cell against cancer. An invasive ductal carcinoma of the breast (pale area at the center) surrounded by spikes of whitish scar tissue and yellow fatty tissue An invasive colorectal carcinoma (top center) in a colectomy specimen A squamous-cell carcinoma (the whitish tumor) near the bronchi in a lung specimen A large invasive ductal carcinoma in a mastectomy specimen Cancer prevention is defined as active measures to decrease the risk of cancer. Dietary recommendations for cancer prevention typically include an emphasis on vegetables, fruit, whole grains, and fish, and an avoidance of processed and red meat (beef, pork, lamb), animal fats, and refined carbohydrates. Carriers of these mutations may then undergo enhanced surveillance, chemoprevention, or preventative surgery to reduce their subsequent risk.  \n",
      "\n",
      "TREATMENT\n",
      "The chance of survival depends on the type of cancer and extent of disease at the start of treatment. Clonal evolution leads to intra- tumour heterogeneity that complicates designing effective treatment strategies. Many treatment options for cancer exist, with the primary ones including surgery , chemotherapy , radiation therapy , hormonal therapy , targeted therapy and palliative care . Which treatments are used depends on the type, location, and grade of the cancer as well as the person's health and wishes. The treatment intent may be curative or not curative. Chemotherapy is the treatment of cancer with one or more cytotoxic anti- neoplastic drugs ( chemotherapeutic agents ) as part of a standardized regimen . The radiation is most commonly low energy x-rays for treating skin cancers while higher energy x-ray beams are used in the treatment of cancers within the body. Surgery is the primary method of treatment of most isolated solid cancers and may play a role in palliation and prolongation of survival. Palliative care refers to treatment that attempts to make the person feel better and may or may not be combined with an attempt to treat the cancer. Unlike treatment that is aimed at directly killing cancer cells, the primary goal of palliative care is to improve the person's quality of life . People at all stages of cancer treatment should have some kind of palliative care to provide comfort. Unlike hospice care, palliative care does not require people to stop treatment aimed at prolonging their lives or curing the cancer. Additionally, an oncologist should consider a palliative care consult in any person they feel has less than months of life even if continuing aggressive treatment. Complementary and alternative cancer treatments are a diverse group of health care systems, practices, and products that are not part of conventional medicine. Some alternative treatments have been investigated and shown to be ineffective but still continue to be marketed and promoted. Taken as a whole, about half of people receiving treatment for invasive cancer (excluding carcinoma in situ and non-melanoma skin cancers) die from cancer or its treatment. The increased risk is believed to be primarily due to the same risk factors that produced the first cancer, partly due to the treatment for the first cancer, and potentially related to better compliance with screening. A centenarian is unlikely to survive for five years even if the treatment is successful. People with lower quality of life may be affected by major depressive disorder and other complications from cancer treatment and/or disease progression that both impairs their quality of life and reduces their quantity of life. BC  AD) translated karkinos into the Latin cancer , also meaning crab and recommended surgery as treatment. The military metaphors also help rationalize radical, destructive treatments. Some psychotherapists said that treatment to change the patient's outlook on life would cure the cancer. The idea was excoriated by the notoriously outspoken Susan Sontag , who published Illness as Metaphor while recovering from treatment for breast cancer in Although the original idea is now generally regarded as nonsense, the idea partly persists in a reduced form with a widespread, but incorrect, belief that deliberately cultivating a habit of positive thinking will increase survival.  including treatment and indirect mortality expenses (such as lost productivity in the workplace)  was estimated to be billion. Angiogenesis inhibitors were once thought to have potential as a \" silver bullet \" treatment applicable to many types of cancer, but this has not been the case in practice. Experimental cancer treatments are treatments that are being studied to see whether they work. Typically, these are studied in clinical trials to compare the proposed treatment to the best existing treatment. They may be entirely new treatments, or they may be treatments that have been used successfully in one type of cancer, and are now being tested to see whether they are effective in another type. More and more, such treatments are being developed alongside companion diagnostic tests to target the right drugs to the right patients, based on their individual biology. In some cases, some or all treatments are postponed until after birth if the cancer is diagnosed late in the pregnancy. Early deliveries to speed the start of treatment are not uncommon. Some treatments, especially certain chemotherapy drugs given during the first trimester , increase the risk of birth defects and pregnancy loss (spontaneous abortions and stillbirths). Some treatments may interfere with the mother's ability to give birth vaginally or to breastfeed her baby. Veterinary oncology , concentrating mainly on cats and dogs, is a growing specialty in wealthy countries, and the major forms of human treatment such as surgery and radiotherapy may be offered.  \n",
      "\n",
      "Prediction on Diabetes_mellitus: Disease \n",
      "\n",
      "Name of the disease: Diabetes mellitus\n",
      "\n",
      "Description of the disease: \n",
      "\n",
      "DRUG\n",
      "Many drugs impair insulin secretion and some toxins damage pancreatic beta cells. Several other groups of drugs, mostly given by mouth, may also decrease blood sugar in type II DM.  \n",
      "\n",
      "SYMPTOM\n",
      "In the United States, diabetes cost billion in The classic symptoms of untreated diabetes are weight loss, polyuria (increased urination), polydipsia (increased thirst), and polyphagia (increased hunger). Several other signs and symptoms can mark the onset of diabetes, although they are not specific to the disease. These typically develop after many years but may be the first symptom in those who have otherwise not received a diagnosis before that time. The symptoms can include numbness, tingling, pain, and altered pain sensation, which can lead to damage to the skin. The rare disease diabetes insipidus has similar symptoms to diabetes mellitus, but without disturbances in the sugar metabolism ( insipidus means \"without taste\" in Latin) and does not involve the same disease mechanisms. This is possibly due to the diet and life-style of the ancient people, or because the clinical symptoms were observed during the advanced stage of the disease. He described the symptoms and the course of the disease, which he attributed to the moisture and coldness, reflecting the beliefs of the \"Pneumatic School\". People with diabetes who have neuropathic symptoms such as numbness or tingling in feet or hands are twice as likely to be unemployed as those without the symptoms. The symptoms may relate to fluid loss and polyuria, but the course may also be insidious.  \n",
      "\n",
      "DIAGNOSIS\n",
      "citation needed Prediabetes indicates a condition that occurs when a person's blood glucose levels are higher than normal but not high enough for a diagnosis of type DM. He hypothesized a correlation of diabetes with other diseases and he discussed differential diagnosis from the snakebite which also provokes excessive thirst.  \n",
      "\n",
      "CAUSE\n",
      "If left untreated, diabetes can cause many complications. Insulin and some oral medications can cause low blood sugar . Prolonged high blood glucose can cause glucose absorption in the lens of the eye , which leads to changes in its shape, resulting in vision changes. Damage to the eyes, known as diabetic retinopathy , is caused by damage to the blood vessels in the retina of the eye, and can result in gradual vision loss and blindness . Additionally, proximal diabetic neuropathy causes painful muscle wasting and weakness. The \"other specific types\" are a collection of a few dozen individual causes. It causes approximately of diabetes mellitus cases in North America and Europe. Type diabetes can affect children or adults, but was traditionally termed \"juvenile diabetes\" because a majority of these diabetes cases were in children. A lack of exercise is believed to cause of cases. Increased levels of insulin in a fetus's blood may inhibit fetal surfactant production and cause respiratory distress syndrome . Some cases of diabetes are caused by the body's tissue receptors not responding to insulin (even when insulin levels are normal, which is what separates it from type diabetes); this form is very uncommon. Any disease that causes extensive damage to the pancreas may lead to diabetes (for example, chronic pancreatitis and cystic fibrosis ). Diseases associated with excessive secretion of insulin-antagonistic hormones can cause diabetes (which is typically resolved once the hormone excess is removed). The following is a comprehensive list of other causes of diabetes: Insulin is the principal hormone that regulates the uptake of glucose from the blood into most cells of the body, especially liver, muscle, and adipose tissue. It is preferable to measure a fasting glucose level because of the ease of measurement and the considerable time commitment of formal glucose tolerance testing, which takes two hours to complete and offers no prognostic advantage over the fasting test. Glycated hemoglobin is better than fasting glucose for determining risks of cardiovascular disease and death from any cause. The World Health Organization (WHO) estimated that diabetes resulted in million deaths in making it the leading cause of death. The discrepancy between the two estimates is due to the fact that cardiovascular diseases are often the cause of death for individuals with diabetes; the IDF uses modelling to estimate the amount of deaths that could be attributed to diabetes.  \n",
      "\n",
      "PREVENTION\n",
      "Limiting sugary beverages and eating less red meat and other sources of saturated fat can also help in the prevention of diabetes.  \n",
      "\n",
      "TREATMENT\n",
      "There are three main types of diabetes mellitus: Prevention and treatment involve a healthy diet , physical exercise , maintaining a normal body weight , and avoiding use of tobacco . Learning about the disease and actively participating in the treatment is important, since complications are far less common and less severe in people who have well-managed blood sugar levels. The goal of treatment is an HbA level of but should not be lower than that, and may be set higher. People with diabetes can benefit from education about the disease and treatment, good nutrition to achieve a normal body weight, and exercise, with the goal of keeping both short-term and long-term blood glucose levels within acceptable bounds . citation needed Metformin is generally recommended as a first line treatment for type diabetes, as there is good evidence that it decreases mortality. Effective treatment was not developed until the early part of the century, when Canadians Frederick Banting and Charles Herbert Best isolated and purified insulin in and This was followed by the development of the long-acting insulin NPH in the The word diabetes ( /  d a . Several countries established more and less successful national diabetes programmes to improve treatment of the disease. The principles of treatment (weight loss, oral antidiabetics, subcutaneous insulin) and management of emergencies (e.g.  \n",
      "\n",
      "CPU times: user 3.11 s, sys: 57.6 ms, total: 3.17 s\n",
      "Wall time: 4.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "kw_list = ['symptom', 'cause', 'prognosis', 'prevention',\n",
    "               'treatment', 'drug', 'susceptibility', 'diagnosis']\n",
    "class_pred = {True: 'Disease', False: 'Non Disease'}\n",
    "\n",
    "urls_test = ['https://en.wikipedia.org/wiki/Paracetamol', 'https://en.wikipedia.org/wiki/Aagenaes_syndrome',\n",
    "        'https://en.wikipedia.org/wiki/Cancer', 'https://en.wikipedia.org/wiki/Diabetes_mellitus']\n",
    "names_test = [u.split('/')[-1] for u in urls_test]\n",
    "\n",
    "for n, url in zip(names_test, urls_test):\n",
    "    y, title, description = classify_from_url(url, count_vect2, tf_transformer2, svm_occ2, kw_list)\n",
    "    print u'Prediction on {}: {} \\n'.format(n, class_pred[y])\n",
    "    if title is not None:\n",
    "        print u'Name of the disease: {}\\n'.format(title)\n",
    "        pretty_print_description(description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
